### PR TITLE
Disable some command when executing graph

### PR DESCRIPTION
**Pull request by ke-yu**
### Purpose

This PR is to partially fix defect [MAGN 7148 Opening Manual execute dyn after running Auto execute dyn twice gives crash](http://adsk-oss.myjetbrains.com/youtrack/issue/MAGN-7148) by disabling some commands when executing graph.

The crash is because while the first execution (of some zero touch function) is still running, open operation tries to reset the VM. To avoid this race condition, here are some solutions, but none of them is satisfying:
  * Kill the current execution. It may cause some resources that allocated in zero touch function not be released properly, so the following Run operations might fail. 
  * Queue the open operation. The problem is, user is not able to know whether open operations are successful or not (no response from UI). And worse, these operations would be handled in short time if each file needs very few time to execute. 
  * Isolate each executions by putting them in different sandboxes. It is the most flexible solution, but requires big change.  

So a workaround is disabling some UI commands. They include
  * Open file
  * Open recent file
  * Create new Home Workspace
  * Close current Home Workspace

### Reviewers

@Benglin  could you please take a look? I don't create unit test because `RaiseCanExecuteChanged` from `RunSettingsViewModel` has to be called from UI thread, but `Application.Current` is null in nunit.

### CODE REVIEW

File: src/DynamoCoreWpf/ViewModels/RunSettingsViewModel.cs [hunk 1]
```diff
@@ -256,6 +256,16 @@ void Model_PropertyChanged(object sender, System.ComponentModel.PropertyChangedE
...
+                    if (Application.Current != null)
+                    {
+                        Application.Current.Dispatcher.Invoke(new Action(() =>
+                        {
+                            dynamoViewModel.ShowOpenDialogAndOpenResultCommand.RaiseCanExecuteChanged();
+                            dynamoViewModel.NewHomeWorkspaceCommand.RaiseCanExecuteChanged();
+                            dynamoViewModel.OpenRecentCommand.RaiseCanExecuteChanged();
+                            dynamoViewModel.CloseHomeWorkspaceCommand.RaiseCanExecuteChanged();
+                        }));
+                    }
```
Comment (Benglin) [inline src/DynamoCoreWpf/ViewModels/RunSettingsViewModel.cs]: Are we saying that `Model_PropertyChanged` might not be called on the main thread?

Comment (ke-yu) [inline src/DynamoCoreWpf/ViewModels/RunSettingsViewModel.cs]: Yes.

Comment (Benglin) [inline src/DynamoCoreWpf/ViewModels/RunSettingsViewModel.cs]: Yep okay that makes sense. We now know `RaisePropertyChanged` automatically handle cross-thread notifications, I was expecting `RaiseCanExecuteChanged` to do the same, but it seems like that isn't the case? Did you explicitly see issue with cross-thread exception if you call `RaiseCanExecuteChanged` directly on different thread? Just being curious...

Comment (ke-yu) [inline src/DynamoCoreWpf/ViewModels/RunSettingsViewModel.cs]: You are absolutely correct - `RaisePropertyChanged` could be called in non-UI thread, but `RaiseCanExecuteChanged` has to be called in UI thread. The reason is, `RaiseCanExecuteChanged` might trigger some UI changes, for example, the enable/disable state of menu item or button, so it has to be called in UI thread (see [this StackOverflow answer](http://stackoverflow.com/questions/15075553/how-can-i-update-a-commands-canexecute-value-from-a-different-thread)).

**Comment by Benglin** [comment]
That makes perfect sense, @ke-yu. Thanks for clarifying (and sorry for getting back to this late)!

**Pull request merged**